### PR TITLE
add luci-theme-kucat Beauty lies in coolness, born from action

### PR DESCRIPTION
**Pull request by sirpdboy**
Maintainer: Sirpdboy team
Development time: December 2021

Main supports LEDE18.06 and the official 18.06 branch.  version: 1.4.9
JS Support Official Website Branch 19.07-23.05   Version: 2.3.9

Function Introduction:

Beauty lies in coolness, born from action
A navigation bar with 5 shortcut keys
Adapt to all versions from 18.06 to 23.05 and optimize most plugins
A large number of optimizations have been made for mobile phones, and there are basically no misalignment or other phenomena
There is a matching theme setting tool that can adjust the light and dark modes, navigation key content, background image, desktop image, and multiple color schemes to choose from.
Integrate light and dark modes.

Of course, there must be some imperfections, and we will continue to update and correct them in the future.

Cool Cat Theme Link: https://github.com/sirpdboy/luci-theme-kucat

kucat2
kucat3
kucat5
kucat6

### CODE REVIEW

**Comment by stokito** [comment]
It looks nice, great job!
The theme uses a template based on lua which was deprecated in the OpenWrt. Now it uses ucode templates. Here is an example https://github.com/openwrt/luci/tree/master/themes/luci-theme-openwrt-2020/ucode/template/themes/openwrt2020

As far I understood you want to support OpenWrt 2018 so the Lua is the only option.
I just want to confirm from you that you are aware about the ucode.

**Review by stokito — APPROVED** [review]

File: themes/luci-theme-kucat/htdocs/luci-static/kucat/css/fonts.css [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+ *      Copyright 2019-2023 sirpdboy <sirpdboy@qq.com>
+ *
+ *  Licensed to the public under the Apache License 2.0
+ */
+@font-face {
+  font-family: 'sirpdboy-kucat';
+  src:  url('../fonts/sirpdboy-kucat.eot?9qz8zf');
+  src:  url('../fonts/sirpdboy-kucat.eot?9qz8zf#iefix') format('embedded-opentype'),
+    url('../fonts/sirpdboy-kucat.ttf?9qz8zf') format('truetype'),
+    url('../fonts/sirpdboy-kucat.woff?9qz8zf') format('woff'),
```
[the remark this answers is not in the document]
Comment (sirpdboy) [reply themes/luci-theme-kucat/htdocs/luci-static/kucat/css/fonts.css:16]: The font library is a dedicated font icon for Kucat, dedicated to this theme, and has no copyright issues.

File: themes/luci-theme-kucat/htdocs/luci-static/kucat/css/fonts.css [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    url('../fonts/sirpdboy-kucat.svg?9qz8zf#sirpdboy-kucat') format('svg');
```
[the remark this answers is not in the document]
Comment (sirpdboy) [reply themes/luci-theme-kucat/htdocs/luci-static/kucat/css/fonts.css:17]: The font library is a dedicated font icon for Kucat, dedicated to this theme, and has no copyright issues.

File: themes/luci-theme-kucat/htdocs/luci-static/kucat/img/logoword [hunk 3]
```diff
@@ -0,0 +1 @@
+做事先做人，尊重他人劳动成果，是为人的基本准则！
```
[the remark this answers is not in the document]
Comment (sirpdboy) [reply themes/luci-theme-kucat/htdocs/luci-static/kucat/img/logoword:1]: This is a preset daily sentence that will be used when there is no internet connection. Avoid affecting aesthetics

File: themes/luci-theme-kucat/htdocs/luci-static/resources/menu-kucat.js [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+		window.addEventListener('resize', this.handleSidebarToggle, true);
+		
+	},
+
+	handleMenuExpand: function(ev) {
+		var a = ev.target, slide = a.parentNode, slide_menu = a.nextElementSibling;
+		var collapse = false;
+
+		document.querySelectorAll('.main .main-left .nav > li >ul.active').forEach(function (ul) {
+			$(ul).stop(true).slideUp("fast", function () {
```
[the remark this answers is not in the document]
Comment (sirpdboy) [reply themes/luci-theme-kucat/htdocs/luci-static/resources/menu-kucat.js:48]: It's okay to change to JS. Is there a plan for improvement?

**Comment by sirpdboy** [comment]
> It looks nice, great job! The theme uses a template based on lua which was deprecated in the OpenWrt. Now it uses ucode templates. Here is an example https://github.com/openwrt/luci/tree/master/themes/luci-theme-openwrt-2020/ucode/template/themes/openwrt2020
> 
> As far I understood you want to support OpenWrt 2018 so the Lua is the only option. I just want to confirm from you that you are aware about the ucode.

The reason for using Lua's template is that this topic supports the latest versions from 19.07 to 23.05. If using the ucode template, only the latest version can be used.

File: themes/luci-theme-kucat/htdocs/luci-static/kucat/css/fonts.css (hunk 1 above)
[the remark this answers is not in the document]
Comment (stokito) [reply themes/luci-theme-kucat/htdocs/luci-static/kucat/css/fonts.css:16]: In the URL `../fonts/sirpdboy-kucat.ttf?9qz8zf` the `?9qz8zf` in the end is q query param. Usualy they are used like this for [cache busting](https://www.keycdn.com/support/what-is-cache-busting).
Maybe you copied this from somewhere and now can remove it.

File: themes/luci-theme-kucat/htdocs/luci-static/resources/menu-kucat.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (stokito) [reply themes/luci-theme-kucat/htdocs/luci-static/resources/menu-kucat.js:48]: Here I found some example how to make the slideUp() in a plain JS:
https://stackoverflow.com/questions/37338030/vanilla-js-version-of-jquery-slidedown-on-dropdown-menu

As for me this is not critical. Maybe someone else later can contribute with improvements.

[the remark this answers is not in the document]
Comment (sirpdboy) [reply themes/luci-theme-kucat/htdocs/luci-static/resources/menu-kucat.js:48]: > Here I found some example how to make the slideUp() in a plain JS: https://stackoverflow.com/questions/37338030/vanilla-js-version-of-jquery-slidedown-on-dropdown-menu
> 
> As for me this is not critical. Maybe someone else later can contribute with improvements.

This example is not a dynamically generated menu, it is a fixed menu item, and this method is not very suitable.

File: themes/luci-theme-kucat/htdocs/luci-static/kucat/css/fonts.css (hunk 1 above)
[the remark this answers is not in the document]
Comment (sirpdboy) [reply themes/luci-theme-kucat/htdocs/luci-static/kucat/css/fonts.css:16]: If this query string is removed, the browser may cache old font files, resulting in updated fonts not displaying correctly. Therefore, it is recommended to keep this query string to ensure that updates to the font file take effect in a timely manner.

[the remark this answers is not in the document]
Comment (stokito) [reply themes/luci-theme-kucat/htdocs/luci-static/kucat/css/fonts.css:16]: 1. Indeed. And we do need the caching.
2. Even if you want to bust cach then you must generate the query string.

FYI the luci has an option to disable cache

**Comment by castillofrancodamian** [comment]
The anime is not suitable.

**Comment by jow-** [comment]
As it stands right now, this PR is a NACK.

There's several points that should be addressed:
 - No new Lua submissions anymore, theme templates have to be in ucode
 - Use the official OpenWrt logo (either https://openwrt.org/_media/logo.png or https://lxr.openwrt.org/openwrt-logo.png)
 - No anime background, rather use something neutral like a simple gradient or pattern
 - Replace cache buster strings with `<%# ?v=PKG_VERSION %>` in templates, see other themes for reference
 - Remove the download functionality and hardcoded curl calls from the theme templates, this should happen in a separate package, providing regular download cronjobs or similar
 - Use only english text and source code comments
 - Ensure that any functionality that fetches external resources is disabled by default
 - Clarify origin and license of the various bundled binary assets (e.g. .woff and .ttf files)
 - Remove usage of jquery.min.js and replace with vanilla .js + CSS animations
 - Rename references to `/etc/config/advancedplus` to something more meaningful, e.g. `/etc/config/kucat` or remove references to that not-existing config file entirely
 - Use proper and consistent indentation
 - Add initramfs warning to theme header

There's likely more which I have overlooked but we can do another review round if the above points have been addressed.

**Pull request closed without merging**